File: api/schemas/cockpit.py
"""JSON models for the cockpit API."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


class KillZone(BaseModel):
    start_iso: str
    end_iso: str


class ScheduleOverlapItem(BaseModel):
    start_iso: str
    end_iso: str
    title_a: str
    title_b: str
    source_a: str
    source_b: str
    id: Optional[str] = None
    start_a_iso: Optional[str] = None
    end_a_iso: Optional[str] = None
    start_b_iso: Optional[str] = None
    end_b_iso: Optional[str] = None


class ScheduleSourceFlag(BaseModel):
    message: str
    start_iso: str
    work_title: str
    personal_title: str
    api_source: str


class SuggestionOption(BaseModel):
    value: str
    label: str


class SuggestionItem(BaseModel):
    id: str
    prompt: str
    options: List[SuggestionOption] = Field(default_factory=list)


class ScheduleDaySignals(BaseModel):
    """Aligned with docs/SCHEDULE_INTAKE_RYAN.md (A1 / A5 / A2 + meeting load + work vs API flags)."""

    overlap_count: int = 0
    overlaps: List[ScheduleOverlapItem] = Field(default_factory=list)
    source_flags: List[ScheduleSourceFlag] = Field(default_factory=list)
    meeting_load_minutes: int = 0
    meeting_load_hours_display: str = "0h"
    meeting_load_warning: bool = False
    meeting_load_warn_threshold_minutes: int = 300
    max_free_gap_minutes: int = 0
    deep_slot_60_available: bool = True
    fragmented_day: bool = False
    small_gap_count: int = 0
    immovable_title_hits: int = 0
    suggestion_questions: List[str] = Field(default_factory=list)
    suggestion_items: List[SuggestionItem] = Field(default_factory=list)
    summary_line: str = ""


class TimedEventItem(BaseModel):
    start_iso: str
    title: str
    source: str  # "google" | "personal" (unchanged for overrides / logic)
    source_kind: Literal["personal_google", "personal_ics", "work_screenshot"] = Field(
        description="personal_* = API feeds; work_screenshot = last Gemini screenshot advisory for this day."
    )
    column_date_iso: Optional[str] = Field(
        default=None,
        description="Printed calendar date for this row's day column (work screenshot / week view).",
    )
    column_weekday: Optional[str] = Field(
        default=None,
        description="English weekday of that column, e.g. Sunday (optional; work screenshot).",
    )


class TacticalBriefLines(BaseModel):
    """Work-calendar Gemini output: three one-line fields (Rugged Executive tone)."""

    fragmentation: str = ""
    kill_zone: str = ""
    priority: str = ""


class TacticalBriefPeriods(BaseModel):
    """Tactical brief split by time-of-day band (screenshot analyze)."""

    morning: TacticalBriefLines = Field(default_factory=TacticalBriefLines)
    afternoon: TacticalBriefLines = Field(default_factory=TacticalBriefLines)
    evening: TacticalBriefLines = Field(default_factory=TacticalBriefLines)


class TomorrowRunwayPreview(BaseModel):
    """Next calendar day runway slice when viewing today's cockpit."""

    date: str = ""
    integrity_wake_iso: Optional[str] = None
    tactical_integrity_wake_iso: Optional[str] = None
    default_wake_iso: str = ""
    runway_conflict: bool = False
    anchor_title: Optional[str] = None
    anchor_start_iso: Optional[str] = None
    anchor_source: Optional[str] = None
    synthetic_default_anchor: bool = False
    prep_shortfall_labels: List[str] = Field(default_factory=list)
    conflict_summary: Optional[str] = None


class GoldenPathTimelineRow(BaseModel):
    start_iso: str
    end_iso: str = ""
    title: str
    source: str = ""
    source_kind: str = ""
    badges: List[str] = Field(default_factory=list)
    expand_hint: str = ""


class GoldenPathProposal(BaseModel):
    id: str
    headline: str
    detail: str
    deltas: Dict[str, Any] = Field(default_factory=dict)
    status: Literal["pending", "approved"] = "pending"


class IntegrityHabitSnapshot(BaseModel):
    posture_sessions_7d: List[bool] = Field(default_factory=list)
    notes: str = ""


class SidebarIntegrity28d(BaseModel):
    """Rolling 28 calendar days ending on recon `date` (posture from protocol history; neck from integrity_stats)."""

    labels: List[str] = Field(default_factory=list, description="Short day labels, oldest → newest.")
    posture_days: List[bool] = Field(default_factory=list)
    neck_days: List[bool] = Field(default_factory=list)


class GraveyardPreviewEntry(BaseModel):
    task_id: str = ""
    title: str = ""
    closed_at: str = ""
    source: str = ""


class WorkCalendarAdvisoryPanel(BaseModel):
    """Last saved work-calendar screenshot analysis for this recon day (if any)."""

    time_coaching: str = ""
    notes: str = ""
    visibility: str = ""
    saved_at: Optional[str] = None
    tactical_brief: Optional[TacticalBriefPeriods] = None


class MorningBriefPayload(BaseModel):
    """EA-style morning optimization scan (deterministic v1)."""

    visible: bool = False
    dismissed: bool = False
    briefing_window_active: bool = False
    anchors_summary: str = ""
    kill_zones_top3: List[KillZone] = Field(default_factory=list)
    trio_slot1_title: str = ""
    matched_zone_index: Optional[int] = None
    brief_markdown: str = ""
    generated_at: str = ""


class SovereigntyPayload(BaseModel):
    """Vanguard Sovereignty Engine headline KPIs."""

    sovereignty_quotient_percent: float = 0.0
    sovereignty_quotient_blended_percent: float = 0.0
    deep_work_sessions_logged: int = 0
    execution_mix_total: int = 0
    utility_tagged_open_count: int = 0
    sovereignty_line: str = ""
    operational_authority_line: str = ""
    financial_sovereignty_line: str = ""
    physical_baseline_line: str = ""


class DeadBugAlert(BaseModel):
    project_id: str = ""
    project_name: str = ""
    hours_since_activity: float = 0.0
    task_id: str = ""
    title_hint: str = ""


class RunwayPayload(BaseModel):
    integrity_wake_iso: Optional[str] = None
    tactical_integrity_wake_iso: Optional[str] = None
    default_wake_iso: str
    runway_conflict: bool
    anchor_title: Optional[str] = None
    anchor_start_iso: Optional[str] = None
    anchor_source: Optional[str] = None
    synthetic_default_anchor: bool = False
    notification_markdown: str
    prep_posture_minutes: int = 0
    prep_neck_minutes: int = 0
    prep_ops_minutes: int = 0
    prep_total_minutes: int = 0
    prep_expected_total_minutes: int = 120
    prep_gap_minutes: int = 0
    prep_expected_posture_minutes: int = 30
    prep_expected_neck_minutes: int = 60
    prep_expected_ops_minutes: int = 30
    prep_shortfall_labels: List[str] = Field(default_factory=list)
    operator_display: str = "You"
    conflict_summary: Optional[str] = None
    tomorrow_preview: Optional[TomorrowRunwayPreview] = Field(
        default=None,
        description="Populated when recon day is today: next-day prep/anchor snapshot.",
    )


class CockpitResponse(BaseModel):
    date: str
    google_calendar_connected: bool = False
    executive_score_percent: float = Field(description="Focus-style score from trio execution counts")
    vanguard_executed: dict = Field(default_factory=dict)
    runway: RunwayPayload
    kill_zones: List[KillZone] = Field(default_factory=list)
    schedule_day_signals: ScheduleDaySignals = Field(default_factory=ScheduleDaySignals)
    daily_landscape: List[TimedEventItem] = Field(default_factory=list)
    personal_calendar_note: str = ""
    personal_calendar_status: Literal["ok", "not_configured", "error"] = "ok"
    identity_alert: bool = False
    integrity_protocol_confirmed: bool = Field(
        default=False,
        description="Physical protocol fully confirmed for calendar *today* when recon day differs, so urgency/red shell matches the real clock; equals recon day when recon is today.",
    )
    work_calendar_advisory: Optional[WorkCalendarAdvisoryPanel] = None
    work_calendar_week_gap_hint: str = Field(
        default="",
        description="Non-empty when a week-style work snapshot exists but no rows match this recon day (usually wrong column_date_iso).",
    )
    execution_day_summary: str = Field(
        default="",
        description="Plain-language merged day summary (landscape + runway + work advisory + schedule read).",
    )
    golden_path_resolution_summary: str = Field(
        default="",
        description="One line built from saved schedule tradeoff MCQs.",
    )
    schedule_tradeoff_answers: Dict[str, str] = Field(default_factory=dict)
    golden_path_proposals: List[GoldenPathProposal] = Field(default_factory=list)
    golden_path_timeline: List[GoldenPathTimelineRow] = Field(default_factory=list)
    golden_path_snoozed: bool = False
    integrity_habit_snapshot: Optional[IntegrityHabitSnapshot] = None
    identity_purpose: str = Field(
        default="",
        description="Life purpose text (identity) for Golden path context; same source as Purpose Pillar.",
    )
    sidebar_integrity: SidebarIntegrity28d = Field(default_factory=SidebarIntegrity28d)
    graveyard_preview: List[GraveyardPreviewEntry] = Field(default_factory=list)
    morning_brief: Optional[MorningBriefPayload] = None
    integrity_consistency_percent: float = Field(
        default=100.0,
        description="Trailing 7d posture/neck + today's protocol blend (discipline visibility).",
    )
    integrity_sentry_state: Literal["NOMINAL", "WARNING", "CRITICAL"] = "NOMINAL"
    ops_posture_nudge_visible: bool = False
    ops_posture_nudge_message: str = ""
    focus_shell_window_active: bool = False
    sacred_integrity_debt_count: int = 0
    cockpit_operator_name: str = Field(
        default="",
        description="COCKPIT_OPERATOR_NAME when set, else runway operator_display.",
    )
    sovereignty: SovereigntyPayload = Field(default_factory=SovereigntyPayload)
    air_gap_active: bool = False
    midday_shield_active: bool = False
    identity_alignment_window_active: bool = False
    air_gap_extension_suggested: bool = Field(
        default=False,
        description="True when prior night sleep below target — advisory longer morning protection.",
    )
    todoist_inbox_open_count: int = 0
    inbox_slaughter_gate_ok: bool = False
    dead_bug_alerts: List[DeadBugAlert] = Field(default_factory=list)
    firefighting_signals: List[str] = Field(default_factory=list)
    firewall_audit_summary: str = ""
    favor_strike_days_clean_7d: int = 0
    favor_strike_streak_7d: int = 0
    commitments_partner_overdue: bool = False
    ragstone_ledger: Dict[str, Any] = Field(default_factory=dict)
    zero_utility_labor_today: bool = False
    evening_wins_count: int = 0
    evening_leaks_count: int = 0

File: web/src/types/cockpit.ts
export type TacticalBriefLines = {
  fragmentation?: string;
  kill_zone?: string;
  priority?: string;
};

export type TacticalBriefPeriods = {
  morning?: TacticalBriefLines;
  afternoon?: TacticalBriefLines;
  evening?: TacticalBriefLines;
};

export type TomorrowRunwayPreview = {
  date: string;
  integrity_wake_iso?: string | null;
  tactical_integrity_wake_iso?: string | null;
  default_wake_iso?: string;
  runway_conflict?: boolean;
  anchor_title?: string | null;
  anchor_start_iso?: string | null;
  anchor_source?: string | null;
  synthetic_default_anchor?: boolean;
  prep_shortfall_labels?: string[];
  conflict_summary?: string | null;
};

export type RunwayPayload = {
  integrity_wake_iso: string | null;
  tactical_integrity_wake_iso: string | null;
  default_wake_iso: string;
  runway_conflict: boolean;
  anchor_title: string | null;
  anchor_start_iso: string | null;
  anchor_source: string | null;
  synthetic_default_anchor: boolean;
  notification_markdown: string;
  prep_posture_minutes?: number;
  prep_neck_minutes?: number;
  prep_ops_minutes?: number;
  prep_total_minutes?: number;
  prep_expected_total_minutes?: number;
  prep_gap_minutes?: number;
  prep_expected_posture_minutes?: number;
  prep_expected_neck_minutes?: number;
  prep_expected_ops_minutes?: number;
  prep_shortfall_labels?: string[];
  operator_display?: string;
  conflict_summary?: string | null;
  tomorrow_preview?: TomorrowRunwayPreview | null;
};

export type LandscapeApiRow = {
  start_iso: string;
  title: string;
  source: string;
  /** Present when API returns extended snapshot; UI falls back from `source` if missing. */
  source_kind?: "personal_google" | "personal_ics" | "work_screenshot";
  /** Printed date on the calendar column (work screenshot / week view). */
  column_date_iso?: string;
  column_weekday?: string;
};

export type ScheduleOverlapItem = {
  start_iso: string;
  end_iso: string;
  title_a: string;
  title_b: string;
  source_a: string;
  source_b: string;
  id?: string;
  start_a_iso?: string;
  end_a_iso?: string;
  start_b_iso?: string;
  end_b_iso?: string;
};

export type ScheduleSourceFlag = {
  message: string;
  start_iso: string;
  work_title: string;
  personal_title: string;
  api_source: string;
};

export type SuggestionOption = { value: string; label: string };

export type SuggestionItem = {
  id: string;
  prompt: string;
  options: SuggestionOption[];
};

/** Intake-aligned schedule pressure (see docs/SCHEDULE_INTAKE_RYAN.md). */
export type ScheduleDaySignals = {
  overlap_count: number;
  overlaps: ScheduleOverlapItem[];
  source_flags: ScheduleSourceFlag[];
  meeting_load_minutes: number;
  meeting_load_hours_display: string;
  meeting_load_warning: boolean;
  meeting_load_warn_threshold_minutes: number;
  max_free_gap_minutes: number;
  deep_slot_60_available: boolean;
  fragmented_day: boolean;
  small_gap_count: number;
  immovable_title_hits: number;
  suggestion_questions: string[];
  suggestion_items?: SuggestionItem[];
  summary_line: string;
};

export type GoldenPathProposal = {
  id: string;
  headline: string;
  detail: string;
  deltas?: Record<string, unknown>;
  status: "pending" | "approved";
};

export type GoldenPathTimelineRow = {
  start_iso: string;
  end_iso?: string;
  title: string;
  source: string;
  source_kind: string;
  badges: string[];
  expand_hint: string;
};

export type IntegrityHabitSnapshot = {
  posture_sessions_7d: boolean[];
  notes: string;
};

export type SidebarIntegrity28d = {
  labels: string[];
  posture_days: boolean[];
  neck_days: boolean[];
};

export type GraveyardPreviewEntry = {
  task_id: string;
  title: string;
  closed_at: string;
  source: string;
};

export type MorningBriefPayload = {
  visible: boolean;
  dismissed: boolean;
  briefing_window_active: boolean;
  anchors_summary: string;
  kill_zones_top3: { start_iso: string; end_iso: string }[];
  trio_slot1_title: string;
  matched_zone_index: number | null;
  brief_markdown: string;
  generated_at: string;
};

export type SovereigntyPayload = {
  sovereignty_quotient_percent?: number;
  sovereignty_quotient_blended_percent?: number;
  deep_work_sessions_logged?: number;
  execution_mix_total?: number;
  utility_tagged_open_count?: number;
  sovereignty_line?: string;
  operational_authority_line?: string;
  financial_sovereignty_line?: string;
  physical_baseline_line?: string;
};

export type DeadBugAlert = {
  project_id?: string;
  project_name?: string;
  hours_since_activity?: number;
  task_id?: string;
  title_hint?: string;
};

export type CockpitPayload = {
  date: string;
  google_calendar_connected: boolean;
  executive_score_percent: number;
  vanguard_executed: { deep: number; mixed: number; shallow: number };
  runway: RunwayPayload;
  kill_zones: { start_iso: string; end_iso: string }[];
  schedule_day_signals?: ScheduleDaySignals;
  daily_landscape: LandscapeApiRow[];
  personal_calendar_note: string;
  identity_alert?: boolean;
  integrity_protocol_confirmed?: boolean;
  personal_calendar_status?: "ok" | "not_configured" | "error";
  /** Last Gemini work-calendar screenshot analysis for this day (server-persisted). */
  work_calendar_advisory?: {
    time_coaching?: string;
    notes?: string;
    visibility?: string;
    saved_at?: string;
    tactical_brief?: TacticalBriefPeriods | null;
  } | null;
  /** Set when week-style work snapshot rows do not match this recon day (re-run analyze). */
  work_calendar_week_gap_hint?: string;
  /** Server-built plain summary: merged landscape + runway + work advisory + schedule read. */
  execution_day_summary?: string;
  golden_path_resolution_summary?: string;
  schedule_tradeoff_answers?: Record<string, string>;
  golden_path_proposals?: GoldenPathProposal[];
  golden_path_timeline?: GoldenPathTimelineRow[];
  golden_path_snoozed?: boolean;
  integrity_habit_snapshot?: IntegrityHabitSnapshot | null;
  identity_purpose?: string;
  sidebar_integrity?: SidebarIntegrity28d;
  graveyard_preview?: GraveyardPreviewEntry[];
  morning_brief?: MorningBriefPayload | null;
  integrity_consistency_percent?: number;
  integrity_sentry_state?: "NOMINAL" | "WARNING" | "CRITICAL";
  ops_posture_nudge_visible?: boolean;
  ops_posture_nudge_message?: string;
  focus_shell_window_active?: boolean;
  sacred_integrity_debt_count?: number;
  cockpit_operator_name?: string;
  sovereignty?: SovereigntyPayload;
  air_gap_active?: boolean;
  midday_shield_active?: boolean;
  identity_alignment_window_active?: boolean;
  air_gap_extension_suggested?: boolean;
  todoist_inbox_open_count?: number;
  inbox_slaughter_gate_ok?: boolean;
  dead_bug_alerts?: DeadBugAlert[];
  firefighting_signals?: string[];
  firewall_audit_summary?: string;
  favor_strike_days_clean_7d?: number;
  favor_strike_streak_7d?: number;
  commitments_partner_overdue?: boolean;
  ragstone_ledger?: Record<string, unknown>;
  zero_utility_labor_today?: boolean;
  evening_wins_count?: number;
  evening_leaks_count?: number;
};

/** Merged Daily landscape row (API + optional session screenshot advisory). */
export type LandscapeRowMerged =
  | LandscapeApiRow
  | {
      start_iso: string;
      title: string;
      source: string;
      source_kind: "work_screenshot";
    };
